### PR TITLE
add defrag command

### PR DESCRIPTION
**Pull request by kellyselden**

### CODE REVIEW

File: bin/commands/defrag.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+module.exports = {
+  command: 'defrag',
+  describe: 'synchronize all dependency version discrepancies',
+  builder: {
```
Comment (NullVoxPopuli) [inline bin/commands/defrag.js:6]: this is very exciting :D 

these options provide much more utility than my one-off script does

File: src/build-dep-graph.js [hunk 2]
```diff
@@ -106,7 +107,9 @@ async function buildDepGraph({
   };
 
   await firstPass(workspaceMeta, workspacePackageJson, packageDirs);
-  secondPass(workspaceMeta);
+  if (shouldPruneDeps) {
+    secondPass(workspaceMeta);
```
Comment (NullVoxPopuli) [inline src/build-dep-graph.js:111]: what does `secondPass` mean? can this be something more meaningful?

File: src/build-dep-graph.js [hunk 3]
```diff
@@ -75,6 +75,7 @@ function secondPass(workspaceMeta) {
 
 async function buildDepGraph({
   workspaceCwd,
+  shouldPruneDeps = true,
```
Comment (NullVoxPopuli) [inline src/build-dep-graph.js:78]: what does it mean to prune deps? based on the added test (which sets the option to false) , I'm not sure

File: src/fs.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    let oldJson = JSON.parse(oldContents);
+
+    let newJson = await callback(oldJson);
+
+    if (!newJson) {
+      // mutation
+      newJson = oldJson;
+    }
+
+    let newContents = JSON.stringify(newJson, null, 2) + EOL;
```
Comment (NullVoxPopuli) [inline src/fs.js:30]: :raised_hands:

File: src/defrag.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+const semverMinVersion = require('semver/ranges/min-version');
+const semverGt = require('semver/functions/gt');
+const semverSatisfies = require('semver/functions/satisfies');
+const semverValidRange = require('semver/ranges/valid');
+const path = require('path');
+const { replaceJsonFile } = require('./fs');
+
+const defaultDependenciesFilter = [];
+
+function firstPass(workspaceMeta, {
```
Comment (NullVoxPopuli) [inline src/defrag.js:18]: maybe: `gatherPackageRanges`?

File: src/defrag.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+        allRanges[packageName].push(versionRange);
+      }
+    }
+  }
+
+  return allRanges;
+}
+
+function secondPass(allRanges, {
```
Comment (NullVoxPopuli) [inline src/defrag.js:44]: maybe: `determineUpdatedPackageRanges`?

File: src/defrag.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  return rangeUpdates;
+}
+
+function thirdPass(workspaceMeta, rangeUpdates, {
```
Comment (NullVoxPopuli) [inline src/defrag.js:80]: maybe: `updatePackageRanges`?

File: src/defrag.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+          console.log(`${_package.packageName}.${dependencyType}.${packageName}: ${oldRange} -> ${newRange}`);
+        } else {
+          dependencies[packageName] = newRange;
+        }
+      }
+    }
+  }
+}
+
+async function fourthPass(workspaceMeta) {
```
Comment (NullVoxPopuli) [inline src/defrag.js:110]: maybe: `writePackageUpdates`?

**Comment by NullVoxPopuli** [comment]
All the tests use the `^` range. for folks who maybe want to pin their deps, how would that work?

File: src/build-dep-graph.js (hunk 3 above)
Comment (kellyselden) [reply src/build-dep-graph.js:78]: before this new feature, monorepo-next was only concerned with packages in the monorepo, and excluded everything else. This brings in the ability to deal with non-monorepo packages for defragging.

**Comment by kellyselden** [comment]
> All the tests use the ^ range. for folks who maybe want to pin their deps, how would that work?

I changed an option to `outOfRange` to deal with pinned deps better and added tests. I think the command you now want is `next defrag --out-of-range minor`.

**Comment by kellyselden** [comment]
I updated the function names to be more descriptive. Thanks for the callout.

**Pull request merged**